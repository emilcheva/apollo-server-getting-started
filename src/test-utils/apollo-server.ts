import { ApolloServer } from 'apollo-server';
import { createTestClient } from 'apollo-server-testing';
import schema from '../schema'; // import existing schema & datasource to utilize the same in test scenarios as well...
import dataSources from '../services'; // services/api's

const createApolloTestServer = () => {
  const server = new ApolloServer({
    schema,
    dataSources,
    debug: true,
    context: () => {
      // Important: request context is not available during jest testing, hence mocked the request context here.
      return { token: 'foo' };
    }
  });
  return server;
};

const apolloServer = createApolloTestServer();

export const { query, mutate } = createTestClient(apolloServer);
