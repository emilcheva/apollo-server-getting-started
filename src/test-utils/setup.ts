import { InMemoryLRUCache } from 'apollo-server-caching';
import BaseService from '../services/Base';

/**
 * Initialize the service.
 * NOTE:
 * ==================================
 * We need to initialize service (dataSource) manually as if it's used through apollo server, apollo server also do the same.
 * ==================================
 * @param Service
 */
export const initialize = <T extends BaseService>(Service: new () => T): T => {
  const service = new Service();

  service.initialize({
    context: {},
    cache: new InMemoryLRUCache()
  });
  return service;
};
