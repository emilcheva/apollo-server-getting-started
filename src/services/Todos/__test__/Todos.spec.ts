import nock from 'nock';
import { initialize } from './../../../test-utils/setup';
import { TodosAPI } from '../index';
import todos from './../../../mocks/todos.json';

describe('Todos RESTDatasource', () => {
  let todosAPI: TodosAPI;
  beforeAll(() => {
    todosAPI = initialize(TodosAPI);
  });

  it('should return todos', async () => {
    nock('https://jsonplaceholder.typicode.com')
      .get('/todos')
      .reply(200, todos);

    const response = await todosAPI.getTodos();
    expect(response).toMatchSnapshot();
  });
});
