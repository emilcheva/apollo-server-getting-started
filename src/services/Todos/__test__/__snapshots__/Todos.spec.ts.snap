// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos Module should return todos 1`] = `
Array [
  Object {
    "completed": false,
    "id": 1,
    "title": "delectus aut autem",
  },
  Object {
    "completed": false,
    "id": 2,
    "title": "quis ut nam facilis et officia qui",
  },
  Object {
    "completed": false,
    "id": 3,
    "title": "fugiat veniam minus",
  },
  Object {
    "completed": true,
    "id": 4,
    "title": "et porro tempora",
  },
  Object {
    "completed": false,
    "id": 5,
    "title": "laboriosam mollitia et enim quasi adipisci quia provident illum",
  },
  Object {
    "completed": false,
    "id": 6,
    "title": "qui ullam ratione quibusdam voluptatem quia omnis",
  },
  Object {
    "completed": false,
    "id": 7,
    "title": "illo expedita consequatur quia in",
  },
  Object {
    "completed": true,
    "id": 8,
    "title": "quo adipisci enim quam ut ab",
  },
  Object {
    "completed": false,
    "id": 9,
    "title": "molestiae perspiciatis ipsa",
  },
  Object {
    "completed": true,
    "id": 10,
    "title": "illo est ratione doloremque quia maiores aut",
  },
]
`;
